### PR TITLE
Add more pre-commit checks (mccabe, bandit, mypy, etc)

### PR DESCRIPTION
**Pull request by djhoese**
Let's bring pyresample into the modern age of 5 years ago with some real complexity and type checking.

 - [ ] Closes #xxxx 
 - [ ] Tests added 
 - [ ] Tests passed 
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
@mraspaud @pnuu I think this is ready. Let me know what you think. I took all of the pre-commit configuration from pre-ruff satpy. The biggest issue/difficulty was mccabe complaining about code complexity which isn't surprising given the age of this package. The biggest "offenders" were the new-ish `area_config.py` stuff which I blame partially on me and my supervision of Will Roberts on adding it. I think this could probably use a rewrite.

The **biggest** problem was the `get_sample_from_neighbour_info` method. This method is...a problem. I got to the point that I couldn't refactor it the way it needed without risking breaking backwards compatibility. Even though we have plenty of tests, I don't trust them to the extent that this method needs to be refactored. As a shortcut/cheat, I mostly just put chunks of code into a function and called it. This meant having to create functions that have a ton of arguments. I think in the long run this functionality will be rewritten in a resampler class in the future and to support DataArrays, dask arrays, and numpy arrays so I'm not tempted to try harder on this. That said, I think here are some avenues that someone could go down if they wanted the code to be better:

1. A helper class: Many of the arguments passed around are kind of like a "state" that is used by multiple areas. A single class with those as instance attributes might make the code look cleaner.
2. Remove all specialization of "multi-channel" arrays and instead always shape input into a 1 or more channel array. Then refactor things into shared functions that don't have to do with the "certainty" or "weighted" functionality so that this function could be split into 3 separate functions (maybe) with each of these 3 cases: nearest, weighted, weighted + certainty.

File: docs/source/conf.py [hunk 1]
```diff
@@ -111,7 +112,7 @@
 
 # List of directories, relative to source directory, that shouldn't be searched
 # for source files.
-exclude_trees = []
+exclude_trees: list[str] = []
```
Comment (mraspaud) [inline docs/source/conf.py:115]: Does mypy really need to run on this file?

**Review by mraspaud — COMMENTED** [review]
Thanks a lot from breaking down all the complexity! I could not check that the logic was preserved in most places, but trust this is mostly refactoring, right?

And how about using ruff in place of flake8, and maybe even bandit?

**Comment by mraspaud** [comment]
Here's the ruff config that would mimic what we have with flake8 today:
```
[tool.ruff]
# See https://docs.astral.sh/ruff/rules/
select = ["E", "W", "B", "D", "T10", "C90",]
line-length = 120

[tool.ruff.pydocstyle]
convention = "google"

[tool.ruff.mccabe]
# Unlike Flake8, default to a complexity level of 10.
max-complexity = 10
```
for bandit, that would be "S", but not I'm unsure the functionality is actually the same.

**Comment by djhoese** [comment]
>  I could not check that the logic was preserved in most places, but trust this is mostly refactoring, right?

Right. There are only two places that I can think of where things were more heavily changed:

1. This is still a refactor, but the legacy area parsing was changed to take advantage of the generator-ness of the file handle. The code is much more readable and tests pass.
2. I changed a warning to an exception in the `area_config.py` parsing of YAML formats where someone types a degrees unit, but likely misspells it. It was detecting "deg" being in units but the overall word was not "degrees" or "deg". I could/should maybe just take that check out completely.

File: docs/source/conf.py (hunk 1 above)
Comment (djhoese) [reply docs/source/conf.py:115]: No, but I also don't mind it. I noticed in a new project that sphinx now removes all unnecessary/optional arguments from `conf.py` when you run the quickstart command. We could consider comparing that to what is in this conf.py and clearing unused things out.

**Comment by djhoese** [comment]
> And how about using ruff in place of flake8, and maybe even bandit?

I had decided not to use it because I did not want to go down the path of all of the rules that I know you worked on for Satpy. The example config you have looks easy enough so I'll try updating to that today.

However, ruff still doesn't handle blank lines between functions/methods and it is pretty annoying to me (it has been annoying to me to review Satpy PRs where there are 3 or more lines between functions). What do you think about waiting until that is handled to switch to ruff?

> for bandit, that would be "S", but not I'm unsure the functionality is actually the same.

How so?

**Comment by mraspaud** [comment]
> However, ruff still doesn't handle blank lines between functions/methods and it is pretty annoying to me (it has been annoying to me to review Satpy PRs where there are 3 or more lines between functions). What do you think about waiting until that is handled to switch to ruff?

oh, ok, I had missed that. That's pretty annoying indeed. We could also have ruff run in parallel with (bare) flake8 then maybe. But otherwise we can wait, yes.
 
> > for bandit, that would be "S", but not I'm unsure the functionality is actually the same.
> 
> How so?

The "S" rule for ruff is an implementation of flake8-bandit, not bandit itself. And flake8-bandit is calling bandit, so I suppose we'd better call bandit directly.

File: .bumpversion.cfg [hunk 2]
```diff
@@ -3,7 +3,7 @@ current_version = 1.10.2a0.dev0
 commit = True
 tag = False
 parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)((?P<pre>[a-z]+)(?P<num>\d+))?(\.(?P<release>dev)(?P<dev>\d+))?
-serialize =
```
Comment (mraspaud) [inline .bumpversion.cfg:6]: Are we still using bumpversion??

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -1,11 +1,40 @@
...
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  rev: 'v0.1.6'
+  hooks:
+  - id: ruff
+# once https://github.com/astral-sh/ruff/issues/2402 is fully resolved then we can get rid of flake8:
+- repo: https://github.com/pycqa/flake8
+  rev: 6.1.0
+  hooks:
+  - id: flake8
+    additional_dependencies: [flake8-docstrings, flake8-debugger, flake8-bugbear, mccabe]
```
Comment (mraspaud) [inline .pre-commit-config.yaml:13]: I think mccabe and maybe bugbear at least are fully covered by ruff, so they could be remove from flake8

**Review by mraspaud — APPROVED** [review]
Looks good, just a couple of comments inline

File: .bumpversion.cfg (hunk 2 above)
Comment (djhoese) [reply .bumpversion.cfg:6]: Haha...uuummm...no idea. I guess I'll say no and remove it.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (djhoese) [reply .pre-commit-config.yaml:13]: Ok. I can remove that.

**Comment by djhoese** [comment]
Ok so CodeFactor is complaining about 11 things which were all existing TODOs in these modules. Codebeat on the other hand says 0 introduced issues and 125 resolved. Not bad.

**Comment by djhoese** [comment]
Merging this now and then I'll hope to resolve conflicts in @ghiggi's PR tonight.

**Pull request merged**